pyomo/contrib/solver/solvers/gurobi/gurobi_direct_base.py

```diff
@@ -0,0 +1,529 @@
+    def release_license():
+        if GurobiDirectBase._gurobipy_env is None:
+            return
+        if GurobiDirectBase._num_gurobipy_env_clients:
+            logger.warning(
+                "Call to GurobiDirectBase.release_license() with %s remaining "
+                "environment clients." % (GurobiDirectBase._num_gurobipy_env_clients,)
+            )
+        GurobiDirectBase._gurobipy_env.close()
+        GurobiDirectBase._gurobipy_env = None
+
+    @staticmethod
+    def env():
+        if GurobiDirectBase._gurobipy_env is None:
+            with capture_output(capture_fd=True):
+                GurobiDirectBase._gurobipy_env = gurobipy.Env()
+        return GurobiDirectBase._gurobipy_env
+
+    @staticmethod
+    def _register_env_client():
+        GurobiDirectBase._num_gurobipy_env_clients += 1
+
+    @staticmethod
+    def _release_env_client():
+        GurobiDirectBase._num_gurobipy_env_clients -= 1
+        if GurobiDirectBase._num_gurobipy_env_clients <= 0:
+            # Note that _num_gurobipy_env_clients should never be <0,
+            # but if it is, release_license will issue a warning (that
+            # we want to know about)
+            GurobiDirectBase.release_license()
+
+    def _check_license(self):
+        try:
+            model = gurobipy.Model(env=self.env())
+        except gurobipy.GurobiError:
+            return Availability.BadLicense
+
+        model.setParam('OutputFlag', 0)
+        try:
+            model.addVars(range(2001))
+            model.optimize()
+            return Availability.FullLicense
+        except gurobipy.GurobiError:
+            return Availability.LimitedLicense
+        finally:
+            model.dispose()
+
+    def version(self):
+        version = (
+            gurobipy.GRB.VERSION_MAJOR,
+            gurobipy.GRB.VERSION_MINOR,
+            gurobipy.GRB.VERSION_TECHNICAL,
+        )
+        return version
+
+    def _create_solver_model(self, pyomo_model):
+        # should return gurobi_model, solution_loader, has_objective
+        raise NotImplementedError('should be implemented by derived classes')
+
+    def _pyomo_gurobi_var_iter(self):
+        # generator of tuples (pyomo_var, gurobi_var)
+        raise NotImplementedError('should be implemented by derived classes')
+
+    def _mipstart(self):
+        for pyomo_var, gurobi_var in self._pyomo_gurobi_var_iter():
+            if pyomo_var.is_integer() and pyomo_var.value is not None:
+                gurobi_var.setAttr('Start', pyomo_var.value)
+
+    def solve(self, model, **kwds) -> Results:
+        start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        orig_config = self.config
+        orig_cwd = os.getcwd()
+        try:
+            config = self.config(value=kwds, preserve_implicit=True)
+
+            # hack to work around legacy solver wrapper __setattr__
+            # otherwise, this would just be self.config = config
+            object.__setattr__(self, 'config', config)
```

Comment:
The whole point of ephemeral options is to *not* change the instance config.  I would strongly recommend passing the local `config` to `_create_solver_model` and `_postsolve`

Reply:
That might work for the direct interfaces, but this way simplifies things tremendously for the persistent interface. This is a temporary change that gets reverted below.